File: 310_hw6.cpp
//MinGW would not recognize .txt files on the command line
//I ended up needed to hardwire some values. Did not try it on a linux
//system where it could've worked without faulter
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
using namespace std;

const int MAX = 30;
void fileOpen(string, fstream& file);
vector<float> fillArr(fstream& file, vector<float>);
vector<float> bubbleSort(vector<float> arr1);
vector<float> selectionSort(vector<float> arr2);
vector<float> merge(vector<float> left, vector<float> right);
void binarySearch(vector<float>, float);

int main(int argc, char *argv[])
{
    vector<float> arr1;
    vector<float> arr2;
    int size1,size2 = 0;
    vector<float> bigArr; 
    string name = ""; 
    fstream file; 
    float target; 

    //cout << "Enter File: ";
    //cin >> name;
    fileOpen("arr1.txt",file);
    cout << "Array 1:" << endl;
    arr1 = fillArr(file, arr1);
    //cout << "Enter File: ";
    //cin >> name;
    fileOpen("arr2.txt",file);
    cout << "Array 2: " << endl;
    arr2 = fillArr(file, arr2); 
    arr1 = bubbleSort(arr1); 
    arr2 = selectionSort(arr2);
    bigArr = merge(arr1, arr2);
    cout << "Target to search for: ";
    cin >> target;
    cout << endl; 
    binarySearch(bigArr, target);
    return 0; 
}
/*
    function to open a file and ensure that it is not empty
    the file stream is a referece that gets used between the fill function and this one
    takes a file name/path as input
*/
void fileOpen(string name, fstream& file){
    file.open(name);
    string num; 
    file >> num;

    if (num == ""){
        cout << "Empty File" << endl;
        file.close();
        exit(-1);
    }
    file.clear();
    file.seekg(0);
    return; 
}

/*
    function to fill the array and make sure that it is full of float/reals
    takes the fstream that was vetted in fileOpen()
    returns the filled vector
*/
vector<float> fillArr(fstream& file, vector<float> arr){
    string num; 
    file >> num;
    while (!file.eof()){
        float elt = stof(num);
        arr.push_back(elt);
        file >> num;
    }
        file >> num;
        float elt = stof(num);
        arr.push_back(elt);
    for(int i = 0; i < arr.size(); i++){
        cout << arr[i] << " "; 
    }
    cout << endl << endl;
    file.close();
    return arr; 
}
/*
    sorts an array with bubble sort. needs a vector
    inputs an unsorted vector and returns a sorted vector
*/
vector<float> bubbleSort(vector<float> arr){
    for(int i = 0; i < arr.size(); i++){
        for (int m = 0; m < arr.size()-i-1; m++){
            if (arr[m] > arr[m+1]){
                swap(arr[m], arr[m+1]);
            }
        }
    }
    cout << "Bubble Sort:" << endl; 
    for (int i = 0; i < arr.size(); i++){
        cout << arr[i] << " ";
    }
    cout << endl << endl; 
    return arr;
}
/*
    sorts and array with selection sort. needs a vector
    inputs an unsorted vector and returns a sorted vector
*/
vector<float> selectionSort(vector<float> arr){
    int minIdx,m = 0; 
    for (int i = 0; i < arr.size(); i++){
        minIdx = i; 
        for (m = i + 1; m < arr.size(); m++){
            if (arr[m] < arr[minIdx]){
                minIdx = m;
            }
        }
        if (minIdx != i){
            swap(arr[minIdx], arr[i]);
        }
    }
    cout << "Selection Sort: " << endl;
    for (int i = 0; i < arr.size(); i++){
        cout << arr[i] << " "; 
    }
    cout << endl << endl; 
    return arr;
}
/*
    merges two arrays and only requires two sorted arrays as input
    returns one merged array sorted
*/
vector<float> merge(vector<float> left, vector<float> right){
    vector<float> arr; 
    int l = 0;
    int r = 0; 
    while ((l < left.size()) && (r < right.size())){
        if (left[l] < right[r]){
            float elt = left[l];
            arr.push_back(elt);
            l++;
        }
        else{
            float elt = right[r];
            arr.push_back(elt);
            r++;
        }
    }
    if (left.size() - l == 0){
        for (r; r < right.size(); r++){
        float elt = right[r];
        arr.push_back(elt); 
        }
    }
    else{
        for (l; l < left.size(); l++){
        float elt = left[l];
        arr.push_back(elt);
        }
    }
    cout << "Merge:" << endl;
    for (int i = 0; i < arr.size(); i++){
        cout << arr[i] << " "; 
    }
    cout << endl << endl; 
    return arr; 
}
/*
    searches through a sorted array for a target
    requires a sorted array and a target
    returns success/failure on the command line
*/
void binarySearch(vector<float> arr, float target){
    int low = 0;
    int high = arr.size() - 1; 
    int mid = (low + high) / 2; 

    while (low < mid){
        if (arr[mid] > target){
            high = mid;
            mid = (low + high) / 2; 
        }
        else if (arr[mid] < target){
            low = mid; 
            mid = (low + high) / 2;
        }
        else
            break; 
    }
    if (arr[mid] == target)
        cout << "Found target @ idx " << mid << "." << endl;
    else
        cout << "Target not found" << endl; 
}

File: 310_hw6.py
import sys


def parse_input(name):
    '''
    function to open a file and then fill a list with the values form that file
    returns the list
    '''
    file = open(name)
    raw = (file.read()).split()
    array = []
    for x in raw:
        array.append(float(x))
    print(array)
    return array

def selection_sort(arr):
    '''
    sorts a list using selection sort
    input is an unsorted list
    outputs a sorted list
    '''
    idx = 0
    while idx < len(arr):
        min_idx = idx
        for i in range(idx,len(arr)):
            if arr[i] < arr[min_idx]:
                min_idx = i
        if idx != min_idx:
            min = arr.pop(min_idx)
            swap = arr.pop(idx)
            arr.insert(idx,min)
            arr.insert(min_idx,swap)
        idx += 1
    print("\nSelection Sort:")
    print(arr) 
    return arr

def bubble_sort(arr):
    '''
    sorts a list using bubble sort
    input is an unsorted list
    outputs a sorted list
    '''
    for i in range(len(arr)):
        for m in range(0, len(arr)-i-1):
            if arr[m] > arr[m+1]:
                arr[m], arr[m+1] = arr[m+1], arr[m]
    print("\nBubble Sort:")  
    print(arr)   
    return arr
def merge(l_arr, r_arr):
    '''
    merges two sorted arrays
    takes two sorted arrays as input
    returns a list in ascending order
    '''
    arr = []
    l, r = 0,0
    while l < len(l_arr)  and r < len(r_arr):
        if l_arr[l] < r_arr[r]:
            arr.append(l_arr[l])
            l += 1
        else:
            arr.append(r_arr[r])
            r += 1
    if len(l_arr) - l == 0:
        for r in range(r, len(r_arr)):
            arr.append(r_arr[r])
    else:
        for l in range(l, len(l_arr)):
            arr.append(l_arr[l])
    print("\nMerge:")
    print(arr)
    return arr

def binary_search(arr, target):
    '''
    searches for a target in a sorted list
    returns result on the command line
    takes a sorted list and a target
    '''
    low, high = 0, len(arr)-1
    mid = (low + high) // 2
    while low < mid:
        if arr[mid] > target:           #lower half
            high = mid
            mid = (low + high) // 2
        elif arr[mid] < target:         #upper half
            low = mid
            mid = (low + high) // 2
        else:                           #on target
            break
    print("\nBianary Search:")
    if arr[mid] == target:
        print("Found target @ ",mid)
    else:
        print("Target not found")

if __name__ == "__main__":
    if len(sys.argv) == 4:
        print("\nArray 1:")
        raw1 = parse_input(sys.argv[1])
        print("\nArray 2:")
        raw2 = parse_input(sys.argv[2])
        arr1 = bubble_sort(raw1)
        arr2 = selection_sort(raw2)     
        big_arr = merge(arr1, arr2)
        binary_search(big_arr, int(sys.argv[3]))